addons/pos_online_payment/controllers/payment_portal.py

```diff
@@ -0,0 +1,299 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import urllib.parse
+
+from odoo import _, http, tools
+from odoo.http import request
+from odoo.exceptions import AccessError, ValidationError, UserError, MissingError
+from odoo.addons.payment.controllers.post_processing import PaymentPostProcessing
+from odoo.addons.payment.controllers import portal as payment_portal
+
+
+class PaymentPortal(payment_portal.PaymentPortal):
+
+    def _check_order_access(self, pos_order_id, access_token):
+        try:
+            order_sudo = self._document_check_access(
+                'pos.order', self._cast_as_int(pos_order_id), access_token)
+        except:
+            raise AccessError(
+                _("The provided order or access token is invalid."))
+
+        if order_sudo.state == "cancel":
+            raise ValidationError(_("The order has been canceled."))
+        return order_sudo
+
+    def _ensure_session_open(self, pos_order_sudo):
+        if pos_order_sudo.session_id.state != 'opened':
+            raise AccessError(_("The POS session is not opened."))
+
+    def _get_partner_sudo(self, user_sudo):
+        partner_sudo = user_sudo.partner_id
+        if not partner_sudo and user_sudo._is_public():
+            partner_sudo = self.env.ref('base.public_user')
+        return partner_sudo
+
+    @staticmethod
+    def _redirect_login(request):
+        return request.redirect(f'/web/login?redirect={urllib.parse.quote(request.httprequest.full_path)}')
+
+    @staticmethod
+    def _is_valid_currency(currency):
+        return currency and currency.active
+
+    @staticmethod
+    def _get_amount_to_pay(order_to_pay_sudo):
+        if order_to_pay_sudo.state in ('paid', 'done', 'invoiced'):
+            return 0.0
+        amount = order_to_pay_sudo._get_checked_next_online_payment_amount()
+        if amount and PaymentPortal._is_valid_amount(amount, order_to_pay_sudo.currency_id):
+            return amount
+        else:
+            return order_to_pay_sudo.get_amount_unpaid()
```

Comment:
You are considering the implementation of the method _is_valid_amount forgiving his role: to guarantee that the amount is valid for the payment portal page.
Whatever the _get_checked_next_online_payment_amount is good or not for the order point of view, it shouldn't, in my opinion, bypass the check of the payment portal.
A payment page is something dangerous in my opinion, I prefer to add more checks than needed, than reducing checks and create bypasses as you suggest.

Reply:
That's your point of view, I have a different one. I prefer to split responsibilities, you prefer to mix everything, allow everyone to change everything and add some checks in some cases only. I consider this dangerous and I try to avoid it.